docs/_extensions/specs.py

```diff
@@ -0,0 +1,72 @@
+import re
+import unittest
+
+from docutils import nodes
+
+
+REFERENCE_RE = re.compile(r'''
+^(
+    (?P<ref>[\w\-]+)
+    |
+    ((?P<link_text>(.+))\s*<(?P<ref_brackets>[\w\-]+)>)
+)$
+''', re.VERBOSE)
+
+SPECS = {
+    'apib': ('API Blueprint', 'https://apiblueprint.org/documentation/specification.html#{anchor}'),
+    'mson': ('MSON', 'https://apiblueprint.org/documentation/mson/specification.html#{anchor}'),
+    'openapi2': ('OpenAPI 2', 'https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#user-content-{anchor}'),
+    'openapi3': ('OpenAPI 3', 'https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#user-content-{anchor}'),
+}
+
+
+# https://docutils.readthedocs.io/en/sphinx-docs/howto/rst-roles.html
+def link(name, rawtext, text, lineno, inliner, options={}, content=[]):
```

Comment:
Not usually a good idea to use `{}` and `[]` as default arguments to a function in Python. Advisable to use `None` and then handle it later in the body of the function:

```python
def foo(content=None):
    ...content or []
```

This is problematic because the the same instance of these instantiated array/dict will be passed as the default argument to link each time, any mutation will apply to further invoations of the function.

Take the following example:

```python
def addHello(content=[]):
    content.append('hello')
    return content
```

```python
>>> addHello()
['hello']
>>> addHello()
['hello', 'hello']
>>> addHello()
['hello', 'hello', 'hello']
```

Reply:
I also realise I think I let this slip through in separate PR in `ghissue`.